flytekit/core/type_engine.py

```diff
@@ -229,16 +230,55 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
         )
 
     def from_binary_idl(self, binary_idl_object: Binary, expected_python_type: Type[T]) -> Optional[T]:
+        """
+        TODO: Add more comments to explain the lifecycle of attribute access.
+        This is for dict, dataclass, and dataclass attribute access.
+        """
         if binary_idl_object.tag == MESSAGEPACK:
             try:
                 decoder = self._msgpack_decoder[expected_python_type]
             except KeyError:
                 decoder = MessagePackDecoder(expected_python_type, pre_decoder_func=_default_msgpack_decoder)
                 self._msgpack_decoder[expected_python_type] = decoder
-            return decoder.decode(binary_idl_object.value)
+            python_val = decoder.decode(binary_idl_object.value)
+            """
+            This is to mock the behavior as none type transformer + union transformer
+
+            @dataclass
+            class DC:
+                a: Optional[int]
+
+            When `a` is None, both MessagePackDecoder[int].decode(a) and MessagePackDecoder[None].decode(a) will be None,
+            which will be ambiguous for the union transformer.
+
+            When `a` is int, both MessagePackDecoder[int].decode(a) and MessagePackDecoder[None].decode(a) will be int,
+            which will be ambiguous for the union transformer.
+
+            This assertion will fail when
+            1. `a` is None and MessagePackDecoder[int].decode(a) is None, which should fail.
+            2. `a` is int and MessagePackDecoder[None].decode(a) is int, which should fail.
+
+            This is to avoid the above ambiguity.
+            """
+            if expected_python_type is NoneType:
+                assert python_val is None
+            else:
+                assert python_val is not None
```

Comment:
Yes, let me try it

Reply:
We have to add these logic in the following place in the Union Transformer, which is too much and make the transformer's code harder to read.

1. https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1904-R1905
2. https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1909-R1910
3. https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1921-R1924

Reply:
And I think the current implementation follows the origin's philosophy, which means
"Handle the function `to python value` in the selected type transformer, succeed or fail is determined by it."
So that's why I do this implementation.

Related Code:

Reply:
okay, that makes sense. thank you

Reply:
Done, thinking how to add tests for this case, extremely important.